apps/runtime/src/main/java/quarks/apps/runtime/MonitorApp.java

```diff
@@ -0,0 +1,204 @@
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonObject;
+
+import quarks.execution.DirectSubmitter;
+import quarks.execution.Job;
+import quarks.execution.services.ControlService;
+import quarks.execution.services.RuntimeServices;
+import quarks.execution.services.job.JobRegistryService;
+import quarks.function.Consumer;
+import quarks.function.Supplier;
+import quarks.runtime.jobregistry.JobEvents;
+import quarks.topology.TStream;
+import quarks.topology.Topology;
+import quarks.topology.TopologyProvider;
+import quarks.topology.mbeans.ApplicationServiceMXBean;
+import quarks.topology.plumbing.PlumbingStreams;
+import quarks.topology.services.ApplicationService;
+
+/**
+ * Job monitoring application.
+ * <p>
+ * The application listens on JobRegistry events and resubmits jobs for which 
+ * an event has been emitted because the job is unhealthy. The monitored 
+ * applications must be registered with an {@code ApplicationService} 
+ * prior to submission, otherwise the monitor application cannot restart 
+ * them.</p>
+ * <p> 
+ * The monitoring application must be submitted within a context which 
+ * provides the following services:
+ * <ul>
+ * <li>ApplicationService - an {@code ApplicationServiceMXBean} control 
+ * registered by this service is used to resubmit failed applications.</li>
+ * <li>ControlService - the application queries this service for an 
+ * {@code ApplicationServiceMXBean} control, which is then used for 
+ * restarting failed applications.</li>
+ * <li>JobRegistryService - generates job monitoring events. </li>
+ * </ul>
+ * </p>
+ */
+public class MonitorApp {
+    private final TopologyProvider provider;
+    private final DirectSubmitter<Topology, Job> submitter;
+    private final Topology topology;
+    private static final Logger logger = LoggerFactory.getLogger(MonitorApp.class);
+
+    /**
+     * Constructs a {@code MonitorApp} with the specified name in the 
+     * context of the specified provider.
+     * 
+     * @param provider the topology provider
+     * @param submitter a {@code DirectSubmitter} which provides required 
+     *      services and submits the application
+     * @param name the application name
+     * 
+     * @throws IllegalArgumentException if the submitter does not provide 
+     *      access to the required services
+     */
+    public MonitorApp(TopologyProvider provider, 
+            DirectSubmitter<Topology, Job> submitter, String name) {
+
+        this.provider = provider;
+        this.submitter = submitter;
+        validateSubmitter();
+        this.topology = declareTopology(name);
+    }
+    
+    /**
+     * Submits the application topology.
+     * 
+     * @return the job.
+     * @throws InterruptedException
+     * @throws ExecutionException
+     */
+    public Job submit() throws InterruptedException, ExecutionException {
```

Comment:
I think this indicates there needs to be some standardization around applications, so they don't each have to implement their own submit method. Thinking that maybe there is an interface applications can implement to provide consistency, and we encourage jobs to start through the ApplicationServiceMXBean, and not by themselves.

I'll put together some proposal, it's wouldn't block this being merged.